File: app/src/main/java/com/mohsenoid/rickandmorty/ui/character/list/CharacterListActivity.java
package com.mohsenoid.rickandmorty.ui.character.list;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;

import androidx.annotation.Nullable;
import androidx.fragment.app.FragmentTransaction;

import com.mohsenoid.rickandmorty.R;
import com.mohsenoid.rickandmorty.injection.DependenciesProvider;
import com.mohsenoid.rickandmorty.ui.base.BaseActivity;

import java.util.ArrayList;
import java.util.List;

public class CharacterListActivity extends BaseActivity {

    static String TAG = CharacterListActivity.class.getSimpleName();
    static String TAG_CHARACTER_LIST_FRAGMENT = "character_list_fragment";

    static String ARG_CHARACTER_IDS = "character_ids";

    private CharacterListFragment characterListFragment;

    public static Intent newIntent(Context context, List<Integer> characterIds) {
        Intent intent = new Intent(context, CharacterListActivity.class);
        intent.putIntegerArrayListExtra(ARG_CHARACTER_IDS, new ArrayList<>(characterIds));
        return intent;
    }

    @Override
    public void injectDependencies(@Nullable Bundle savedInstanceState, DependenciesProvider dependenciesProvider) {
        if (savedInstanceState == null) {
            List<Integer> characterIds = getIntent().getIntegerArrayListExtra(ARG_CHARACTER_IDS);
            if (characterIds == null) {
                onBackPressed();
                return;
            }

            characterListFragment = dependenciesProvider.getCharacterListFragment(characterIds);
        } else {
            characterListFragment = (CharacterListFragment) getSupportFragmentManager().findFragmentByTag(TAG_CHARACTER_LIST_FRAGMENT);
        }
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        if (null == savedInstanceState) {
            attachFragments();
        }

        Log.i(TAG, "CharacterListActivity created");
    }

    private void attachFragments() {
        FragmentTransaction fragmentTransaction = getSupportFragmentManager().beginTransaction();
        fragmentTransaction.replace(R.id.container, characterListFragment, TAG_CHARACTER_LIST_FRAGMENT);
        fragmentTransaction.commitAllowingStateLoss();
    }
}

File: app/src/main/kotlin/com/mohsenoid/rickandmorty/view/character/list/CharacterListActivity.kt
package com.mohsenoid.rickandmorty.view.character.list

import android.content.Context
import android.content.Intent
import android.os.Bundle
import com.mohsenoid.rickandmorty.R
import com.mohsenoid.rickandmorty.view.base.BaseActivity
import timber.log.Timber
import java.util.ArrayList

class CharacterListActivity : BaseActivity() {

    private lateinit var characterListFragment: CharacterListFragment

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_character_list)

        val characterIds: List<Int>? = intent.getIntegerArrayListExtra(ARG_CHARACTER_IDS)
        if (characterIds == null) {
            onBackPressed()
            return
        }

        characterListFragment =
            supportFragmentManager.findFragmentByTag(TAG_CHARACTER_LIST_FRAGMENT) as? CharacterListFragment
                ?: CharacterListFragment.newInstance(characterIds)

        if (savedInstanceState == null) {
            attachFragments()
        }
        Timber.i("CharacterListActivity created")
    }

    private fun attachFragments() {
        supportFragmentManager.beginTransaction().apply {
            replace(R.id.container, characterListFragment, TAG_CHARACTER_LIST_FRAGMENT)
        }.commit()
    }

    companion object {
        private const val TAG_CHARACTER_LIST_FRAGMENT: String = "character_list_fragment"

        private const val ARG_CHARACTER_IDS: String = "character_ids"

        fun newIntent(context: Context?, characterIds: List<Int>): Intent {
            return Intent(context, CharacterListActivity::class.java).apply {
                putIntegerArrayListExtra(ARG_CHARACTER_IDS, ArrayList(characterIds))
            }
        }
    }
}
